[SourceKit] Determine correctly whether typed  variables in `if/guard/while-let`-statements have explicit type annotations

fwcd opened the pull request:
As noted in https://github.com/apple/sourcekit-lsp/pull/408#issuecomment-867802689, `if/guard/while-let`s are currently (incorrectly) always reported as having no explicit type annotation as the `TypedPattern` is not the top-level-pattern in such declarations. Consider the following example:

```swift
if let x: Int = Optional.some(4) {}
```

Here we get the following AST:

```
(if_stmt ...
  (pattern
    (pattern_optional_some implicit type='Int?'  <--- this node prevents the current mechanism from discovering the type annotation
      (pattern_typed type='Int'
        (pattern_let implicit type='Int'
          (pattern_named type='Int' 'x'))
        (type_ident
          (component id='Int' bind=Swift.(file).Int))))
    (call_expr ...))
  (brace_stmt ...))
```

`sourcekitd-test` therefore reports `(explicit type: 0)` despite an explicit annotation being present:

```
<VariableTypes>
(1:8, 1:9): Int (explicit type: 0)
</VariableTypes>
```

This PR fixes the issue by including an implicitly generated `OptionalSomePattern` as a special case in `VariableTypeCollector` and also adds tests to verify that these cases are handled correctly.

cc @ahoppen

fwcd committed:
Add VariableType test cases for if/guard/while-let

- Add VariableType test case for guarded variables
- Add if-let to VariableType test case
- Add while-let test case for VariableType
- Test pattern matching with VariableType
- Test guard/while-case-let with VariableType

fwcd committed:
Handle typed if/case/while-let in VariableTypeCollector

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -808,6 +808,20 @@ class VariableTypeCollector : public SourceEntityWalker {
+      if (!HasExplicitType) {
+        if (auto *somePattern =
+                dyn_cast_or_null<OptionalSomePattern>(VD->getParentPattern())) {
+          if (somePattern->isImplicit()) {
```
Is it strictly necessary to check if the `OptionalSomePattern` is implicit here?

fwcd replied:
I don't think that having a typed pattern in an explicit pattern is possible in Swift currently, I included this check just to be sure that we consider the expected case only though.

ahoppen reviewed (commented):
I think it might be worth adding the check to `getTypeReprOrParentPatternTypeRepr`. It seems generic enough to me that other user’s of the API could also benefit from it.

Or was there a specific reason why you didn’t do that?

fwcd commented:
It would change the semantics of the method slightly (by not just considering the parent pattern), therefore I wasn't sure whether this should be included.

ahoppen commented:
I think changing the semantics of `getTypeReprOrParentPatternTypeRepr` is fine if it doesn’t break any tests – I would consider it an improvement to that method. 

I would also leave the implicitness check unless there’s a specific reason to include it that you can explain. Just makes the code simpler and more generic.

fwcd committed:
Move if-let special case into getTypeReprOrParentPatternTypeRepr

As suggested during PR review.

This mechanism seems to be general enough to fit into
getTypeReprOrParentPatternRepr.

ahoppen commented:
@swift-ci Please smoke test

fwcd commented:
Not entirely sure about the CI failure, but it seems unrelated to the PR?

ahoppen commented:
Yes, the CI failure is unrelated. I’ll try again later.

ahoppen commented:
@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please smoke test Linux
